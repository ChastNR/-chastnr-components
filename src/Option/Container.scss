.slt__opts {
  background-color: #fff;
  border-radius: 4px;
  border: 1px solid #eaecee;
  box-shadow: 0 8px 16px rgba(51, 51, 51, 0.2);
  cursor: default;
  height: auto;
  max-height: 300px;
  margin: 0 -16px;
  opacity: 0;
  overflow: hidden;
  position: absolute;
  top: 57px;
  transition: opacity 0.2s ease-out;
  width: calc(100% + 2px);
  z-index: -1;
}
