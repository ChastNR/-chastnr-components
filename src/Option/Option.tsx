import { useCallback, useState } from "react";

import Checkbox from "../Checkbox";
import { Option } from "../types";
import { mouseDownPrevent } from "../utils";

interface OptionProps {
  option: Option;
  checked?: boolean;
  isMulti?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
}

const SelectOption: React.FC<OptionProps> = ({ checked = false, option, isMulti, onClick }) => {
  const [isChecked, setChecked] = useState(checked);

  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLButtonElement>) => {
      setChecked((s) => !s);

      if (onClick) {
        onClick(e);
      }
    },
    [onClick]
  );

  return (
    <button className="slt__opt" onClick={handleClick} onMouseDown={mouseDownPrevent} type="button">
      <div className="slt__opt__txt">{option.title}</div>
      {isMulti && <Checkbox checked={isChecked} />}
    </button>
  );
};

export default SelectOption;
