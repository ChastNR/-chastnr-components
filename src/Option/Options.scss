.slt__opts {
  background-color: #fff;
  border-radius: 4px;
  border: 1px solid #eaecee;
  box-shadow: 0 8px 16px rgba(51, 51, 51, 0.2);
  cursor: default;
  height: auto;
  margin: 0 -16px;
  opacity: 0;
  overflow: hidden;
  position: absolute;
  top: 57px;
  transition: opacity 0.2s ease-out;
  width: calc(100% + 2px);
  z-index: -1;

  .slt__opt {
    align-items: flex-start;
    background-color: #fff;
    color: rgba(0, 0, 0, 0.8);
    cursor: pointer;
    display: flex;
    flex-direction: row;
    min-height: 44px;
    padding-bottom: 10px;
    padding-left: 16px;
    padding-right: 12px;
    padding-top: 10px;
    width: 100%;
    text-align: left;
    border: 0;

    &:hover {
      background-color: #f6f7f8;
    }

    .slt__opt__txt {
      align-items: stretch;
      align-self: stretch;
      display: flex;
      flex-direction: column;
      flex: 1 1 auto;
      justify-content: center;
      overflow: hidden;

      & > div {
        font-size: 15px;
        word-wrap: break-word;
      }
    }

    .cb {
      margin-left: 0;
      margin-right: 12px;
      order: -1;
    }
  }
}
