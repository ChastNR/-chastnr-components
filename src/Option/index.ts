export { Options } from './Options';
export { OptionsContainer } from './Container';
export { SelectOption } from './Option';
