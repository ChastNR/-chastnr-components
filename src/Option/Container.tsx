import { PropsWithChildren } from 'react';
import './Container.scss';

export const OptionsContainer: React.FC<PropsWithChildren<{}>> = ({ children }) => (
  <div className="slt__opts">{children}</div>
);
