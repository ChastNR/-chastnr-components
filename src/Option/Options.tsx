import { Option } from "../types";
import OptionsContainer from "./Container";
import SelectOption from "./Option";

import "./Options.scss";

interface OptionsProps {
  options: Option[];
  isMulti?: boolean;
  onClick: (value: Option) => () => void;
}

const Options: React.FC<OptionsProps> = ({ options, onClick, isMulti = false }) => (
  <OptionsContainer>
    {options.map((o) => (
      <SelectOption isMulti={isMulti} key={o.value} onClick={onClick(o)} option={o} />
    ))}
  </OptionsContainer>
);

export default Options;
