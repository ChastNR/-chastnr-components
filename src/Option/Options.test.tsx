import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Option } from './Option';
import Options from './Options';

describe('<Options />', () => {
  const options: Option[] = [
    {
      title: 'Test',
      value: 'testValue',
    },
    {
      title: 'Test 2',
      value: 'testValue2',
    },
  ];

  const onClickMock = jest.fn();

  it('shoud equal snapshot', () => {
    const optionsComponent = shallow(<Options onClick={onClickMock} options={options} />);
    expect(toJson(optionsComponent)).toMatchSnapshot();
  });

  it('should render options', () => {
    const optionsComponent = mount(<Options onClick={onClickMock} options={options} />);
    const optionComponents = optionsComponent.find('button');
    expect(optionComponents.length).toBeGreaterThan(1);
  });

  it('should render options with checkboxes', () => {
    const optionsComponent = mount(<Options isMulti onClick={onClickMock} options={options} />);
    const inputComponents = optionsComponent.find('input');
    expect(inputComponents.length).toBeGreaterThan(1);
  });
});
