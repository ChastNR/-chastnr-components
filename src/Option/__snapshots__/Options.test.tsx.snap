// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Options /> shoud equal snapshot 1`] = `
<div
  class="slt__opts"
>
  <button
    class="slt__opt"
    type="button"
  >
    <div
      class="slt__opt__txt"
    >
      Test
    </div>
  </button>
  <button
    class="slt__opt"
    type="button"
  >
    <div
      class="slt__opt__txt"
    >
      Test 2
    </div>
  </button>
</div>
`;
